lib/elasticity/index_mapper.rb

```diff
@@ -144,21 +144,23 @@ def map_hit(hit)
       attrs.merge!(hit["_source"]) if hit["_source"]
 
       if hit["highlight"]
-        highlighted_attrs = attrs.dup
-        attrs_set = Set.new
+        hit["highlight"]["name.foo"] = ["o"]
 
-        hit["highlight"].each do |name, v|
+        highlighted_attrs = attrs.dup
+        ordered_set = hit["highlight"].each_with_object([]) do |(name, value), set|
           name = name.gsub(/\..*\z/, '')
-          next if attrs_set.include?(name)
-          highlighted_attrs[name] = v
-          attrs_set << name
+          next if set.any? { |item| item.has_key?(name.to_sym) }
+
+          set.push({ "#{name}": value })
+          highlighted_attrs[name] = value
```

Comment:
This looks more complex than needed. The values should be accessed through the highlighted object, the only thing we need is the order of the highlights which should only contains the ordered list of the attribute names.

I think the simpler way to solve the issue is to simply change `attrs_set` to an array:

```ruby
if hit["highlight"]
  highlighted_attrs = attrs.dup
  highlighted_attr_names = []  # this used to be attrs_set = Set.new

  hit["highlight"].each do |name, v|
    name = name.gsub(/\..*\z/, '')
    next if highlighted_attr_names.include?(name)
    highlighted_attrs[name] = v
    highlighted_attr_names << name
  end

  highlighted = @document_klass.new(highlighted_attrs)
end

injected_attrs = attrs.merge({
  highlighted: highlighted,
  highlighted_attrs: highlighted_attr_names, # injects this 
  _explanation: hit["_explanation"]
})
```

Then you have:

```ruby
result = MyDocument.search(...).first
result.highlighted_attrs # => ["foo", "bar"]
result.highlighted.foo
result.highlighted.bar
```

We don't want to introduce a way to read the highlighted value other than through the highlighted object.